memcache/selector_test.go

```diff
@@ -16,24 +16,284 @@ limitations under the License.
 
 package memcache
 
-import "testing"
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net"
+	"os/exec"
+	"sync"
+	"testing"
+	"time"
+)
 
-func BenchmarkPickServer(b *testing.B) {
-	// at least two to avoid 0 and 1 special cases:
-	benchPickServer(b, "127.0.0.1:1234", "127.0.0.1:1235")
+func TestAllocations(t *testing.T) {
+	t.Run("no allocations with two servers", func(t *testing.T) {
+		benchPickServer(t, "127.0.0.1:1234", "127.0.0.1:1235")
+	})
+	t.Run("no allocations with one server", func(t *testing.T) {
+		benchPickServer(t, "127.0.0.1:1234")
+	})
 }
 
-func BenchmarkPickServer_Single(b *testing.B) {
-	benchPickServer(b, "127.0.0.1:1234")
+
+func benchPickServer(t *testing.T, servers ...string) {
+	br := testing.Benchmark(func(b *testing.B) {
+		b.ReportAllocs()
+		var ss ServerList
+		requireNoError(b, ss.SetServers(servers...))
+		for i := 0; i < b.N; i++ {
+			if _, err := ss.PickServer("some key"); err != nil {
+				b.Fatal(err)
+			}
+		}
+	})
+	if br.AllocsPerOp() > 0 {
+		t.Errorf("expected no allocations, got %d", br.AllocsPerOp())
+	}
+}
+
+// Note: this test is technically non-deterministic as the retry scheduling is async,
+// but failures should be exceedingly rare due to the length of time until the first timeout.
+//
+// If a fix is required: probably simplest to change the startingWait to a var and modify in this test
+func TestBehaviourSynchronously(t *testing.T) {
+	create := func(t *testing.T) (ServerSelector, net.Addr, net.Addr) {
+		ss := createWithServers(t, []string{"127.0.0.1:4200", "127.255.0.1:4200"})
+		addrA := ss.addrs[0]
+		addrB := ss.addrs[1]
+		if addrA == addrB {
+			t.Fatalf("for test to be valid addresses must differ")
+		}
+		return ss, addrA, addrB
+	}
+
+	t.Run("after a net error, an address should be not be pickable until retried", func(t *testing.T) {
+		ss, addrA, _ := create(t)
+
+		ss.OnResult(addrA, &net.OpError{})
+		fuzzer(func(k string) {
+			addr, err := ss.PickServer(k)
+			requireNoError(t, err)
+			if addrA == addr {
+				t.Fatalf("address still available")
+			}
+		})
+	})
+
+	t.Run("after a successful retry, an address should be available again", func(t *testing.T) {
```

Comment:
I think the test case description doesn't totally match what is happening, because the test just reports a `nil` error for `addrA` (but there was no previous known error) -- so the test verifies that on the happy path, when no error ever occurs, that `PickServer` behaves as expected.

Reply:
Good spot, thanks! This test was split up into three and wasn't updated to reflect that the previous test wasn't affecting the same client.